File: cpp/ibahdapi_sampleclient.cpp
#include <iostream>

#include <grpcpp\grpcpp.h>
#include <ibaHD-API.grpc.pb.h>

#include <memory>
#include <string>
#include <sstream>
#include <fstream>
#include "ApiKeyAuthenticator.h"

using namespace grpc;
using namespace ibahdapi::v1;

void read_file(const std::string& filename, std::string& data)
{
	std::ifstream file(filename.c_str(), std::ios::in);
	if (file.is_open())
	{
		std::stringstream ss;
		ss << file.rdbuf();
		file.close();
		data = ss.str();
	}
	return;
}

int main(int argc, char* argv[]) {
	const std::string server = "127.0.0.1:9003";
	const std::string apiKey = "<< ADD API KEY >>";
	const std::string certPath = "<< ADD PATH TO EXPORTED SERVER CERTIFICATE >>";

	// Read exported (PEM formatted) server TLS certificate from disk
	std::string cert;
	read_file(certPath, cert);
	grpc::SslCredentialsOptions opts = { cert };
	auto tls_cert_credentials = grpc::SslCredentials(opts);

	// Apply the api key to every request that is made with the client
	auto api_key_credentials = grpc::MetadataCredentialsFromPlugin(
		std::unique_ptr<grpc::MetadataCredentialsPlugin>(new ApiKeyAuthenticator(apiKey)));

	auto combined_credentials = grpc::CompositeChannelCredentials(tls_cert_credentials, api_key_credentials);

    // Increasing the default message size (~4MB) is recommended
	ChannelArguments channel_args;
	channel_args.SetMaxReceiveMessageSize(INT_MAX);

    // gRPC channel setup to connect to ibaHD-API endpoint in ibaHD-Server
	auto channel = grpc::CreateCustomChannel(server, combined_credentials, channel_args);

	// Instantiate ibaHD-API client on the gRPC channel
	auto client = HdApiService::NewStub(channel);

	// Simple request to retrieve stores
	ClientContext context;

	auto response = GetHdStoresResponse();
	auto status = client->GetHdStores(&context, GetHdStoresRequest(), &response);

	if (!status.ok())
	{
		std::cout << status.error_message() << std::endl;
	}

	for (auto hdstore : response.hd_stores()) {

		std::cout << hdstore.hd_store_name() << std::endl;
	}

	printf("\n\tPress Enter to exit...\n");
	std::cin.get();
}

File: python/ibahdapi_sampleclient.py
from __future__ import print_function
import logging
import grpc
import ibaHD_API_pb2
import ibaHD_API_pb2_grpc

# Helper class to apply the api key to each call
class ApiKeyCallCredentials(grpc.AuthMetadataPlugin):

    def __init__(self, apikey):

        self._apikey = apikey

    def __call__(self, context, callback):

        metadata = (('ibahdapi-apikey', self._apikey),)

        callback(metadata, None)

def run():

    # Read exported (PEM formatted) server TLS certificate from disk
    certificate = open('<< ADD PATH TO EXPORTED SERVER CERTIFICATE >>', 'rb').read()
    tls_credentials = grpc.ssl_channel_credentials(certificate)

    # Apply the api key to every request that is made with the client
    apikey_credentials  = grpc.metadata_call_credentials(ApiKeyCallCredentials('<< ADD API KEY >>'))

    combined_credentials = grpc.composite_channel_credentials(tls_credentials, apikey_credentials)

    # gRPC channel setup to connect to ibaHD-API endpoint in ibaHD-Server
    # Increasing the default message size (~4MB) is recommended (c int32 max = 2147483647)
    with grpc.secure_channel('127.0.0.1:9003', combined_credentials, options=[('grpc.max_receive_message_length', 2147483647)]) as channel:

        # Instantiate ibaHD-API client on the gRPC channel
        client = ibaHD_API_pb2_grpc.HdApiServiceStub(channel)

        # Simple request to retrieve stores
        response = client.GetHdStores(request=ibaHD_API_pb2.GetHdStoresRequest())

        # Clean channel shutdown
        channel.close()

        print(response)

if __name__ == '__main__':
    logging.basicConfig()
    run()
